src/pages/markdown/2022-dont-rerender-all-flatlist-items/index.md

```diff
@@ -0,0 +1,294 @@
+  },
+  {
+    "id": "1",
+    "value": 6,
+    "question": "Menu easy to read"
+  },
+  {
+    "id": "2",
+    "value": 9,
+    "question": "Order was taken promptly"
+  },
+]
+```
+
+State in _React_ is _immutable_, which means we can't update a single item in an array and use the same array again to render the next UI.
+
+If we want to change the rate `value` in a single answer, let's say "_Menu easy to read_" from _6_ to _7_, we have to create a copy of the old list with a single item different. It results in a new `answers` object.
+
+In console logs "_Data has changed_" represents `answers` value change, but we see it each time user updates just one of his answers. It's because whenever user slides one list item slider new `answers` list is passed as a new `data` prop value, which results in re-rendering the whole `FlatList` and all it's children. 
+
+However, there is a way to avoid those unnecessary re-renders and save CPU time. It will help slow smartphones to catch up with performance. 🐌
+
+## Let's give data the shape
+
+What if we could change the shape of the `data` list prop passed to `FlatList` to have the same value always?
+
+In the list `data`, we can store just ids referring to "full" items. It's just a `string[]` list, not a list of complex objects with a `value` field. When the user slides the slider and updates the `value,` it won't change the item `id`. The `data` list will stay unchanged.
+
+This will be passed as a `data` prop:
+
+```javascript
+const answersIdsList = [
+  "0",
+  "1",
+  "2",
+]
+```
+
+Since now data stored in `state` has this shape:
+
+```javascript
+{
+  dataIdsList: ["0", "1", "2"], // answersIdsList
+  dataById: {
+    "0": {
+      "value": 8,
+      "question": "Wide selection in the menu"
+    },
+    "1": {
+      "value": 6,
+      "question": "Menu easy to read"
+    },
+    // ...
+  };
+}
+```
+
+To get the list of items without knowing too many details about item internals, we can use `dataIdsList`. This object won't be updated when _the details_ of some items change, which won't cause re-rendering.
+
+To get details of a single object, we can use `dataById`. For example, the updated list item component (`NormalizedFormListItem`) uses it to get item details while having only `id` as an input: `state.dataById[id]`.
+
+Old `renderItem` needed a full `answer` object as a param. It can be simplified:
+
+```javascript
+// before fixes
+const renderItem: ListRenderItem<SurveyAnswer> = useCallback(
+  ({ item: answer }) => {
+    const setValue = (newValue: number) => {
+      updateAnswer({ id: answer.id, value: newValue });
+    };
+
+    return <FormListItem answer={answer} setValue={setValue} />;
+  },
+  []
+);
+```
+
+
+The new `renderItem` that supports normalized data now as a param needs only item id (`answerId`) to render complete `NormalizedFormListItem`. There is also no need for `useCallback` anymore because we can now move it out of the `HomeScreen` component:
+
+```javascript
```

Comment:
I see you're using javascript code block here.
I remember I was using ```tsx and ```ts, but I don't know if that makes any difference. You can check that, if you're bored :)

Reply:
great work btw.

Reply:
Thanks! It makes a difference. Fixed 👍